Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All). Check your output to make sure it all looks as you expected.

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name below:

In [197]:
NAME = "DUCROCQ Romain"

---

You should have read all the readings and content from the lectures given on Brightspace.


In order to make it easy to grade your works, here are the notes you should read and follow carrefully.  We will take off points for not following them. We will grade based on assuming you have read and understood them.
    
1. **Using the Notebook to Show Your Work**: You must learn to write code in the notebook... It is a core tool for data science and will make it easier to develop and document your work if you become good at using it. Writing your code in another tool and pasting it into the notebook will probably not work well: Some of you forgot to include code elsewhere, or the spacing became messed up, or it didn't run as copied.  You must be sure your code cells execute because we will test them.  So learn how to run code in the notebook cells to double check your work.
2. **Read Directions Carefully**:  The instructions for your code are very important. If you don't follow the requirements, your application won't do as requested. Making it work correctly is part of learning to program.  If we worded something unclearly, send an e-mail to your professor.

---

# Homework Questions:

## Working on lists and dictionaries (and still loops, ...)

**Question **:

Create a function that takes a file name (and path if needed) as the argument.  In the function, open and read in the file **mountains.csv**. Use a try/except to be sure the file exists and is readable. If the file location is wrong or it can't be opened, print an error that begins with "Error:".  (You can test it with a junk path or filename that doesn't exist.)

The pattern I suggest for this is:

````
try:
    with open('mountains.csv', 'r') as handle:
        for line in handle:
        #....do stuff here (you can have other try/except in here if you want)
except:
    print("Error: Something wrong with your file location?")
    return
````

(This pattern is recommended as it doesn't need files to be closed after.)

Mountains.csv is a comma-separated list of mountains, their height in meters, and the range they belong to. (Look at it in a text editor, but don't edit/modify the file!)  A CSV file is a common format for raw data.  Other types of raw data files are point-virgule (semi-colon) separated files or tab-separated files.  However the columns are separated, you must use that character in your "split" code.

In this case, it's a comma. Split each line by the comma, and make a dictionary where the key is the mountain name (the first element) and the height is the value, the second element. Make sure to convert the height to a number.
Then **print the keys and values of the dictionary using ``.items()``, in readable sentences** that say, for instance, "The height of K2 is 8611 meters."  **Return the dictionary at the end of the function.**

Reminder about print with {} in your string: use print(string.format(variable)) to fill in the {} with your variable. If there are 2 {}'s, use `.format(var1, var2)`.



**Note:**
You can use multiple functions if you want, but you have to have one main function that calls the others, if you do. This is because the test code relies on the main function to check your output.



In [198]:
def mountain_height(filename):
    """ Read in a csv file of mountain names and heights.  
    Parse the lines and print the names and heights. 
    Return the data as a dictionary. 
    The key is the mountain and the height is the value.
    """

    mountains = dict()
    msg = "The height of {} is {} meters."
    err_msg = "Error: File doesn't exist or is unreadable."
    
    try:
        with open(filename, 'r') as handle:
            for line in handle:
                mountains[line.split(",")[0]] = int(line.split(",")[1])
    except:
        print("Error: Something wrong with your file location?")
        return None
    
    for key, value in mountains.items():
        print(msg.format(key, value))
        
    return mountains

In [199]:
# Define here the path of your file. So you can use the test after
# Do not forgot to execute this cell before the test ones
filename = 'mountains.csv'

In [200]:
# Use the definition of your file mountains.csv above.  
# Show that it runs.

mountain_height(filename)

The height of Mount Everest is 8848 meters.
The height of K2 is 8611 meters.
The height of Kangchenjunga is 8586 meters.
The height of Lhotse is 8516 meters.
The height of Makalu is 8485 meters.
The height of Cho Oyu is 8201 meters.
The height of Dhaulagiri is 8167 meters.
The height of Manaslu is 8163 meters.
The height of Nanga Parbat is 8126 meters.
The height of Annapurna is 8091 meters.
The height of Gasherbrum I is 8080 meters.
The height of Broad Peak is 8051 meters.
The height of Gasherbrum II is 8035 meters.
The height of Shishapangma is 8027 meters.


{'Mount Everest': 8848,
 'K2': 8611,
 'Kangchenjunga': 8586,
 'Lhotse': 8516,
 'Makalu': 8485,
 'Cho Oyu': 8201,
 'Dhaulagiri': 8167,
 'Manaslu': 8163,
 'Nanga Parbat': 8126,
 'Annapurna': 8091,
 'Gasherbrum I': 8080,
 'Broad Peak': 8051,
 'Gasherbrum II': 8035,
 'Shishapangma': 8027}

In [201]:
# Test code for grading your function. You can ignore this.

output = mountain_height(filename)
assert len(output.keys()) == 14
assert output['Annapurna'] == 8091

The height of Mount Everest is 8848 meters.
The height of K2 is 8611 meters.
The height of Kangchenjunga is 8586 meters.
The height of Lhotse is 8516 meters.
The height of Makalu is 8485 meters.
The height of Cho Oyu is 8201 meters.
The height of Dhaulagiri is 8167 meters.
The height of Manaslu is 8163 meters.
The height of Nanga Parbat is 8126 meters.
The height of Annapurna is 8091 meters.
The height of Gasherbrum I is 8080 meters.
The height of Broad Peak is 8051 meters.
The height of Gasherbrum II is 8035 meters.
The height of Shishapangma is 8027 meters.


In [202]:
# Test code for your printing the data.

from io import StringIO
import mock
import sys

with mock.patch('sys.stdout', new_callable=StringIO):
    mountain_height(filename)
    assert "8848" in sys.stdout.getvalue()
    assert "Mount Everest" in sys.stdout.getvalue()

In [203]:
# Test for your error condition with a bad filename/path.

from io import StringIO
import mock
import sys

filename = None
with mock.patch('sys.stdout', new_callable=StringIO):
    mountain_height(filename)
    assert "Error:" in sys.stdout.getvalue()

---

**Question 2**:

Rewrite the function about exploring the mountains file we worked on during the practice. 
Use the collections module's Counter to count how many times each mountain range is mentioned. Each row contains a mountain, its height, and the range it is part of. The ranges are still in the 3rd column of the mountains.csv file!  You can use more than one function if you want.

Also add a dictionary that records all the heights of the mountains in a particular range. You will use a list for the values of the heights. So this is a dictionary with a list value! The key will be the range name. Each time you see a new mountain in the range, add the height to the list for that key. For example, after reading all the data, ``mountains['Himalayas'] == [8848, 8586, 8516, 8485, 8201, 8167, 8163, 8126, 8091, 8027]``.  (The "Himalayas" are the range.)

You may use a regular ``dict`` or a ``defaultdict``, but you must beware of ``KeyError`` with a regular dictionary if the key doesn't exist yet.

Your output should be to print the top 2 ranges (according to their Counter value -- hint: look at the function ``most_common()``). And adding the mountain range name to the counter requires a little care (look at update).

Then, print the average height of the mountains in each range. (They don't have to be in order. Hint: You may need to find out how to import a ``mean`` function, or else calculate it by hand.)

Return the dictionary object with the ranges and their lists of mountain heights after all the printing.

Show that this code works with the other file, "highest_mountains.csv" too.


In [204]:
# Using Counter()

from collections import Counter
from collections import defaultdict
from statistics import mean   # this also exists in numpy if you prefer

# define your dicts inside the function, so they can be re-used each time it is called.
def mountain_ranges(filename):
    ranges = []
    heights = defaultdict(list)  # empty list is the default here, not an int!
    
    err_msg = 'Error: Check the file path or content'
    
    try:
        with open(filename, 'r') as handle:
            for line in handle:
                heights[line.split(",")[2]].append(int(line.split(",")[1]))
    except:
        print(err_msg)
        return None
    
    ranges = [c[0] for c in Counter(heights).most_common(2)]
    for r in ranges:
        print(mean(heights[r]))
        print(len(heights[r]))
        
    return heights

In [205]:
# Give here the path of your mountains file
filename = 'mountains.csv'

In [206]:
mountain_ranges(filename)

8321
10
8194.25
4


defaultdict(list,
            {'Himalayas': [8848,
              8586,
              8516,
              8485,
              8201,
              8167,
              8163,
              8126,
              8091,
              8027],
             'Karakoram': [8611, 8080, 8051, 8035]})

In [207]:
## Testing the output contains values we expect from the counts and means.

import mock
from io import StringIO
import sys

with mock.patch('sys.stdout', new_callable=StringIO):
    mountain_ranges(filename)
    assert "8321" in sys.stdout.getvalue()
    assert "10" in sys.stdout.getvalue()

In [208]:
# Testing your output for the grade. Ignore this.  Handgrading of the printed output.

result = mountain_ranges(filename)
assert result['Karakoram'] == [8611, 8080, 8051, 8035]

8321
10
8194.25
4


In [209]:
#File path for the highest_mountains file
filepath_highest="highest_mountains.csv"

In [210]:
# Show your code works with the other file, highest_mountains too. 
mountain_ranges(filepath_highest)


7866.083333333333
12
7820
10


defaultdict(list,
            {'Mahalangur Himalaya\n': [8848,
              8516,
              8485,
              8188,
              7952,
              7864,
              7804,
              7543,
              7350,
              7321,
              7309,
              7213],
             'Baltoro Karakoram\n': [8611,
              8080,
              8051,
              8034,
              7946,
              7932,
              7545,
              7410,
              7315,
              7276],
             'Kangchenjunga Himalaya\n': [8586, 7711, 7462, 7412, 7362, 7350],
             'Dhaulagiri Himalaya\n': [8167,
              7751,
              7661,
              7618,
              7385,
              7268,
              7246],
             'Manaslu Himalaya\n': [8163, 7893, 7871],
             'Nanga Parbat Himalaya\n': [8126],
             'Annapurna Himalaya\n': [8091, 7937, 7555, 7455, 7219],
             'Jugal Himalaya\n': [8027],
             'Hispar Karakoram\n':

---

Here is the last exercise of this assignment.

This is a very specific function to be programmed, but it will allow you to train on using loops, lists and dictionary.

We want to create a function that takes one argument which is a CSV file (here it only works with SpaceMission_data.csv file). 

The function explores the file and returns a dictionary in which it counts :
- The number of mission for each company (company will be keys of the dictionary)
- The number of successful missions 
- The number of failed missions

Each value element of the dictionary will be a list of the 3 counts described above. 
It will look like this: 
{company : [total number of missions, number of successful missions, number of failed missions]}

Explore the file first, and find a solution to identify the successful and failed missions identification. 

You will see that it is not only a question about a "," delimiter. 
So in order to handle easily the reading of the format, i strongly suggest you to use this method: 
Just after opening the file, 
use "content = csv.reader(handle)"

And you will be able to loop into content.

And of course return a dictionary! 

In [211]:
import csv
def space_statistics(filename):
    '''This function takes one argument which should be a CSV file. 
    It specifically explores the space_corrected.csv file. 
    The function returns a dictionary in which it counts the number of mission for each company in charge
    and also counts the number of successful and failed missions'''

    space = dict()
    err_Msg = 'Error: Check the file path or content' 
    
    
    try:
        with open(filename, 'r', errors='ignore') as csvfile:
            data = csv.reader(csvfile, delimiter=',')
            next(data, None)
            for row in data:
                if not row[2] in space.keys():
                    space[row[2]] = [0, 0, 0]

                space[row[2]] = [space[row[2]][0] + 1, 
                                 space[row[2]][1] + ("Success" in row[-1]),
                                 space[row[2]][2] + ("Failure" in row[-1])]        
    except:
        print(err_Msg)
        return None

    return space

In [212]:
# Specify here the path of your SpaceMission_data file
file = 'SpaceMission_data.csv'

In [213]:
space_statistics(file)

{'SpaceX': [100, 94, 6],
 'CASC': [251, 231, 20],
 'Roscosmos': [55, 51, 4],
 'ULA': [140, 139, 1],
 'JAXA': [7, 6, 1],
 'Northrop': [83, 74, 9],
 'ExPace': [10, 9, 1],
 'IAI': [11, 9, 2],
 'Rocket Lab': [13, 11, 2],
 'Virgin Orbit': [1, 0, 1],
 'VKS RF': [201, 188, 13],
 'MHI': [84, 80, 4],
 'IRGC': [1, 1, 0],
 'Arianespace': [279, 269, 10],
 'ISA': [13, 4, 9],
 'Blue Origin': [12, 12, 0],
 'ISRO': [76, 63, 13],
 'Exos': [4, 0, 4],
 'ILS': [46, 45, 1],
 'i-Space': [1, 1, 0],
 'OneSpace': [1, 0, 1],
 'Landspace': [1, 0, 1],
 'Eurockot': [13, 12, 1],
 'Land Launch': [7, 6, 1],
 'CASIC': [5, 3, 2],
 'KCST': [5, 2, 3],
 'Sandia': [1, 0, 1],
 'Kosmotras': [22, 21, 1],
 'Khrunichev': [1, 1, 0],
 'Sea Launch': [36, 33, 3],
 'KARI': [3, 1, 2],
 'ESA': [13, 9, 4],
 'NASA': [203, 186, 17],
 'Boeing': [136, 131, 5],
 'ISAS': [30, 26, 4],
 'SRC': [3, 2, 1],
 'MITT': [7, 6, 1],
 'Lockheed': [79, 74, 5],
 'AEB': [3, 0, 3],
 'Starsem': [1, 1, 0],
 'RVSN USSR': [1777, 1614, 163],
 'EER': [1, 0, 1],
 

In [214]:
# Testing the output 
import mock
from io import StringIO
import sys

space = {'SpaceX': [100, 94, 6], 'CASC': [251, 231, 20], 'Roscosmos': [55, 51, 4], 'ULA': [140, 139, 1], 'JAXA': [7, 6, 1], 'Northrop': [83, 74, 9], 'ExPace': [10, 9, 1], 'IAI': [11, 9, 2], 'Rocket Lab': [13, 11, 2], 'Virgin Orbit': [1, 0, 1], 'VKS RF': [201, 188, 13], 'MHI': [84, 80, 4], 'IRGC': [1, 1, 0], 'Arianespace': [279, 269, 10], 'ISA': [13, 4, 9], 'Blue Origin': [12, 12, 0], 'ISRO': [76, 63, 13], 'Exos': [4, 0, 4], 'ILS': [46, 45, 1], 'i-Space': [1, 1, 0], 'OneSpace': [1, 0, 1], 'Landspace': [1, 0, 1], 'Eurockot': [13, 12, 1], 'Land Launch': [7, 6, 1], 'CASIC': [5, 3, 2], 'KCST': [5, 2, 3], 'Sandia': [1, 0, 1], 'Kosmotras': [22, 21, 1], 'Khrunichev': [1, 1, 0], 'Sea Launch': [36, 33, 3], 'KARI': [3, 1, 2], 'ESA': [13, 9, 4], 'NASA': [203, 186, 17], 'Boeing': [136, 131, 5], 'ISAS': [30, 26, 4], 'SRC': [3, 2, 1], 'MITT': [7, 6, 1], 'Lockheed': [79, 74, 5], 'AEB': [3, 0, 3], 'Starsem': [1, 1, 0], 'RVSN USSR': [1777, 1614, 163], 'EER': [1, 0, 1], 'General Dynamics': [251, 203, 48], 'Martin Marietta': [114, 100, 14], 'Yuzhmash': [2, 2, 0], 'Douglas': [1, 1, 0], 'ASI': [9, 9, 0], 'US Air Force': [161, 129, 32], 'CNES': [8, 6, 2], 'CECLES': [4, 1, 3], 'RAE': [2, 1, 1], 'UT': [5, 1, 4], 'OKB-586': [2, 2, 0], 'AMBA': [8, 4, 4], "Arm??e de l'Air": [4, 3, 1], 'US Navy': [17, 2, 15]}

with mock.patch('sys.stdout', new_callable=StringIO):
    assert space_statistics(file) == space


In [215]:
# Some test to verify if all lines have been counted
import mock
from io import StringIO
import sys

with mock.patch('sys.stdout', new_callable=StringIO):
    assert sum([space_statistics(file)[keys][0] for keys in space]) == 4324


In [216]:
# Some test to verify one result from the output dictionary
import mock
from io import StringIO
import sys

with mock.patch('sys.stdout', new_callable=StringIO):
    assert space_statistics(file)['SpaceX'] == [100,94,6]


## Congrats! You have done some datascience with Python !
You can also have alook at PANDAS (https://pandas.pydata.org). 
This Python librairy is mostly used in data science to explore, and play with data (from files, servers, or else).

For your information, The **MK course, Python data analysis** is made for that at emlyon. 